src/rust/engine/rule_graph/src/lib.rs

```diff
@@ -786,31 +793,85 @@ pub fn params_str<T: TypeId>(params: &ParamTypes<T>) -> String {
   T::display(params.iter().cloned())
 }
 
-///
-/// TODO: Move all of these methods to Display impls.
-///
 pub fn entry_str<R: Rule>(entry: &Entry<R>) -> String {
-  match entry {
-    Entry::WithDeps(ref e) => entry_with_deps_str(e),
-    Entry::Param(type_id) => format!("Param({})", type_id),
+  entry_node_str_with_attrs(entry).entry_str
+}
+
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub struct GraphVizEntryWithAttrs {
+  entry_str: String,
+  attrs_str: Option<String>,
+}
+
+pub enum Palette {
+  Olive,
+  Gray,
+  Orange,
+  Blue,
+}
+
+impl Palette {
+  // https://c.eev.ee/kouyou/ is wonderful for selecting lovely color juxtapositions across multiple
+  // different color axes.
+  fn color_string(&self) -> String {
+    // These color values are all in HSV. See https://www.graphviz.org/doc/info/colors.html for
+    // other methods of specifying
+    // colors. https://renenyffenegger.ch/notes/tools/Graphviz/attributes/_color/index may also be
+    // useful.
+    match self {
+      Self::Olive => "0.2214,0.7179,0.8528".to_string(),
+      Self::Gray => "0.576,0,0.6242".to_string(),
+      Self::Orange => "0.08,0.5,0.976".to_string(),
+      Self::Blue => "0.5,1,0.9".to_string(),
+    }
+  }
+
+  pub fn color_format_string(&self) -> String {
```

Comment:
I think `impl fmt::Display for Palette` would make this enum nicer to use (mode idiomatic). You could then replace `Palette::Olive.color_format_string()` with `fmt!("{}", Palette::Olive)`.

Reply:
Because I wanted to maintain the same distinction we have between `fmt::Display` for `Rule` and its `String` representation for the graph visualization, I've instead extracted this method out to `fmt_for_graph()` on the new trait `DisplayForGraph`. I'm not sure how easy it is to introduce new `format()` strings, but that could be the next step!